File: engine/scoring.py
"""Scoring helpers for end-of-turn and final-score calculations."""

from __future__ import annotations

from engine.state import CardDefinition, GameState, NodeKind, _cow_player, board_index, card_index


def _count_troops(node_troop_slots: list[str | None], player_id: str) -> int:
    return sum(1 for occupant in node_troop_slots if occupant == player_id)


def _site_control_owner(state: GameState, node_id: str) -> str | None:
    node_state = state.board.nodes[node_id]
    counts: dict[str, int] = {}

    for occupant in node_state.troop_slots:
        if occupant is not None:
            counts[occupant] = counts.get(occupant, 0) + 1

    if not counts:
        return None

    max_count = max(counts.values())
    leaders = [owner for owner, count in counts.items() if count == max_count]
    return leaders[0] if len(leaders) == 1 and leaders[0] != "white" else None


def _is_total_control(state: GameState, node_id: str, player_id: str) -> bool:
    node_state = state.board.nodes[node_id]
    has_enemy_spies = bool(node_state.spies - {player_id})
    all_troops_owned = all(slot_owner == player_id for slot_owner in node_state.troop_slots)
    return all_troops_owned and not has_enemy_spies


def _cards_vp(cards: list[str], index: dict[str, CardDefinition], field_name: str) -> int:
    total = 0
    for card_id in cards:
        card = index.get(card_id)
        if card is None:
            continue
        total += getattr(card, field_name)
    return total


def award_end_of_turn_site_vp(state: GameState, player_id: str) -> GameState:
    """Add end-of-turn VP from sites where the player has total control."""

    updated = state._cow_clone()
    board_definition_index = board_index(updated.definition.board)

    site_vp = 0
    for node_id, node_definition in board_definition_index.items():
        if node_definition.kind != NodeKind.SITE:
            continue

        if _is_total_control(updated, node_id, player_id):
            site_vp += node_definition.total_control_vp_per_turn

    _cow_player(updated, player_id).score += site_vp
    return updated


def compute_final_scores(state: GameState) -> dict[str, int]:
    """Compute final VP tally from map, deck, inner-circle, trophies, and tokens."""

    definition_by_node = board_index(state.definition.board)
    card_by_id = card_index(state.definition.catalog)
    totals: dict[str, int] = {}

    for player_id, player_state in state.players.items():
        total_vp = player_state.score

        for node_id, node_definition in definition_by_node.items():
            if node_definition.kind != NodeKind.SITE:
                continue

            if _site_control_owner(state, node_id) == player_id:
                total_vp += node_definition.control_vp

        trophy_vp = len(player_state.trophy_hall)
        token_vp = player_state.vp_tokens
        deck_zone_cards = player_state.deck + player_state.hand + player_state.discard_pile
        deck_vp = _cards_vp(deck_zone_cards, card_by_id, "deck_vp")
        inner_circle_vp = _cards_vp(player_state.inner_circle, card_by_id, "inner_circle_vp")

        totals[player_id] = (
            total_vp
            + trophy_vp
            + token_vp
            + deck_vp
            + inner_circle_vp
        )

    return totals

File: engine_c/scoring.c
#include "scoring.h"
#include <string.h>

static int count_troops(const NodeState *ns, Sym player_id) {
    int count = 0;
    for (int i = 0; i < ns->troop_slot_count; i++) {
        if (ns->troop_slots[i] == player_id) count++;
    }
    return count;
}

int site_control_owner(const GameState *state, Sym node_id) {
    for (int i = 0; i < state->node_count; i++) {
        if (state->nodes[i].node_id != node_id) continue;
        const NodeState *ns = &state->nodes[i];

        int counts[MAX_PLAYERS + 1] = {0};
        Sym owners[MAX_PLAYERS + 1];
        int owner_count = 0;

        for (int j = 0; j < ns->troop_slot_count; j++) {
            Sym occ = ns->troop_slots[j];
            if (occ == SYM_NULL) continue;
            int found = -1;
            for (int k = 0; k < owner_count; k++) {
                if (owners[k] == occ) { found = k; break; }
            }
            if (found >= 0) counts[found]++;
            else { owners[owner_count] = occ; counts[owner_count] = 1; owner_count++; }
        }
        if (owner_count == 0) return -1;

        int max_count = 0, max_idx = -1, tie = 0;
        for (int k = 0; k < owner_count; k++) {
            if (counts[k] > max_count) { max_count = counts[k]; max_idx = k; tie = 0; }
            else if (counts[k] == max_count) { tie = 1; }
        }
        if (tie || max_idx < 0) return -1;
        Sym leader = owners[max_idx];
        if (intern_str(leader) && strcmp(intern_str(leader), "white") == 0) return -1;
        return max_idx >= 0 ? (int)leader : -1;
    }
    return -1;
}

int is_total_control(const GameState *state, Sym node_id, Sym player_id) {
    for (int i = 0; i < state->node_count; i++) {
        if (state->nodes[i].node_id != node_id) continue;
        const NodeState *ns = &state->nodes[i];
        for (int j = 0; j < ns->troop_slot_count; j++) {
            if (ns->troop_slots[j] != player_id) return 0;
        }
        for (int j = 0; j < ns->spy_count; j++) {
            if (ns->spies[j] != player_id) return 0;
        }
        return 1;
    }
    return 0;
}

static int cards_vp(const GameState *state, Sym *cards, int count, int field_is_deck_vp) {
    int total = 0;
    for (int i = 0; i < count; i++) {
        for (int j = 0; j < state->definition->catalog.card_count; j++) {
            const CardDefinition *cd = &state->definition->catalog.cards[j];
            if (cd->card_id == cards[i]) {
                total += field_is_deck_vp ? cd->deck_vp : cd->inner_circle_vp;
                break;
            }
        }
    }
    return total;
}

void award_end_of_turn_site_vp(GameState *state, Sym player_id) {
    int site_vp = 0;
    for (int i = 0; i < state->definition->board.node_count; i++) {
        const NodeDefinition *nd = &state->definition->board.nodes[i];
        if (strcmp(intern_str(nd->kind), "site") != 0) continue;
        if (is_total_control(state, nd->node_id, player_id)) {
            site_vp += nd->total_control_vp_per_turn;
        }
    }
    if (site_vp > 0) {
        PlayerState *ps = cow_player(state, player_id);
        if (ps) ps->score += site_vp;
    }
}

void compute_final_scores(const GameState *state, int *scores_out) {
    for (int p = 0; p < state->player_count; p++) {
        Sym pid = state->players[p].player_id;
        const PlayerState *ps = &state->players[p];
        int total = ps->score;

        for (int i = 0; i < state->definition->board.node_count; i++) {
            const NodeDefinition *nd = &state->definition->board.nodes[i];
            if (strcmp(intern_str(nd->kind), "site") != 0) continue;
            int owner = site_control_owner(state, nd->node_id);
            if (owner >= 0 && (Sym)owner == pid) {
                total += nd->control_vp;
            }
        }

        total += ps->trophy_hall_count;
        total += ps->vp_tokens;

        int all_cards[MAX_ZONE_SIZE * 3];
        int all_count = 0;
        for (int j = 0; j < ps->deck_count; j++) all_cards[all_count++] = ps->deck[j];
        for (int j = 0; j < ps->hand_count; j++) all_cards[all_count++] = ps->hand[j];
        for (int j = 0; j < ps->discard_pile_count; j++) all_cards[all_count++] = ps->discard_pile[j];
        total += cards_vp(state, all_cards, all_count, 1);

        total += cards_vp(state, ps->inner_circle, ps->inner_circle_count, 0);

        scores_out[p] = total;
    }
}

Sym find_winner(const GameState *state) {
    int scores[MAX_PLAYERS];
    compute_final_scores(state, scores);

    int max_score = -1, max_idx = -1, tie = 0;
    for (int i = 0; i < state->player_count; i++) {
        if (scores[i] > max_score) { max_score = scores[i]; max_idx = i; tie = 0; }
        else if (scores[i] == max_score) { tie = 1; }
    }
    if (tie || max_idx < 0) return SYM_NULL;
    return state->players[max_idx].player_id;
}
